### PR TITLE
Cancel payments on stock shortages

### LINKED ISSUE #7922: Payment attempts on orders with stock restrictions appear as authorized and "uncaptured" on Stripe
## Description

Related to #7692.
Follow-up issue from #7779 - This PR introduces a warning to the user about an unsuccessful payment when stock is reduced, when the payment awaits for SCA authentication; it also removes doubling of transaction fees, on every payment attempt.

However, for that scenario, he payment still goes through as authorized and authenticated, appearing as "uncaptured" on Stripe.

Both payments appear on Stripe, (1) and (2):

- The first payment attempt (1) above appears in Stripe as an authorized, but `Uncaptured` payment;
- The second and successful payment attempt appears in Stripe as a `Succeeded`  payment;

image

## Expected Behavior

Payment attempts on orders with stock restrictions should be voided, when stock restrictions occur.

## Actual Behaviour

Payment attempts on orders with stock restrictions appear as authorized and "uncaptured" on Stripe

## Steps to Reproduce

As a customer:
1. Add stock to cart
2. Checkout with Stripe SCA
3. Wait till the "Authorization screen" is displayed on the UI

As admin:
4. Set available stock to zero, for that item

As a customer
5. Click `Complete Authentication` (customer)
6. Check the Stripe Dashboard for that account - see the payment is appearing as uncaptured

## Animated Gif/Screenshot

Steps to reproduce:
Peek 2021-07-13 12-52

Payment, on Stripe Dashboard:
image

## Workaround

Cancelling the payment on Stripe:
image

## Severity

bug-s3: a feature is broken but there is a workaround

## Your Environment

* Version used: v3.8.4
* Browser name and version: Firefox 89
* Operating System and version (desktop or mobile): Ubuntu 20.04

### PR DESCRIPTION
**Pull request by Matt-Yorkley**
#### What? Why?

Closes #7922 

Payments should now be correctly voided via `ActiveMerchant` when checkout fails due to stock shortages.

#### What should we test?

Payments with Stripe should be cancelled in Stripe when checkout fails because of stock running out during checkout.

#### Release notes

Fixed issue with payments when checkout fails due to stock shortages.

Changelog Category: Technical changes

### CODE REVIEW

**Review by mkllnk — APPROVED** [review]
Good to have a quick fix. :rocket: 

A spec would be nice for this. It would need some Stripe API interaction though which would lead us to VCR again.

**Review by andrewpbrett — APPROVED** [review]

**Comment by filipefurtad0** [comment]
Hey @Matt-Yorkley, 

This is a tricky one: It seems these changes now lead to an error 500, when in the case of stock restriction, while awaiting for authorization:
Peek 2021-07-29 15-52

As above, the payment is incomplete:
image

But after granting authorization it moves to `uncaptured`:
image

I've checked, that we still have the previously reported behavior in master:

Peek 2021-07-29 16-05

**Comment by Matt-Yorkley** [comment]
There was a bit of a mess here that meant the `#void` method for StripeSCA wasn't actually voiding payments :see_no_evil:  I think I've got it fixed up now, but I'll need to check it in staging a bit.

File: app/models/spree/gateway/stripe_sca.rb [hunk 1]
```diff
@@ -78,7 +82,13 @@ def void(response_code, _creditcard, gateway_options)
...
                                                                  stripe_account: stripe_account_id)
         gateway_options[:stripe_account] = stripe_account_id
-        provider.refund(refundable_amount(payment_intent_response), response_code, gateway_options)
+
+        # If a payment has been confirmed it cannot be voided by Stripe, and must be refunded instead
+        if voidable?(payment_intent_response)
+          provider.void(response_code, gateway_options)
+        else
+          provider.refund(refundable_amount(payment_intent_response), response_code, gateway_options)
+        end
```
Comment (Matt-Yorkley) [inline app/models/spree/gateway/stripe_sca.rb:91]: Re-review note: this was the weird bit. 
- `StripeSCA#purchase` calls `provider#purchase` :heavy_check_mark: 
- `StripeSCA#capture` calls `provider#capture` :heavy_check_mark: 
- `StripeSCA#void` calls... `provider#refund`! What!?

Comment (Matt-Yorkley) [reply app/models/spree/gateway/stripe_sca.rb:91]: This was changed a while ago. The issue is that we can't void (cancel) payments that are confirmed/completed. But that change meant that we also couldn't void incomplete payments at all :see_no_evil:

Comment (Matt-Yorkley) [reply app/models/spree/gateway/stripe_sca.rb:91]: Documentation for reference: https://stripe.com/docs/api/payment_intents/cancel (the void method via `ActiveMerchant` calls this endpoint)

**Comment by RachL** [comment]
Alright, here is the test I've done:

- shopper A goes to spaghetti farm and uses a card that needs authorization and buy big spaghettiiiiii. They buy 1 and there is 1 left. They use a card with authorization, but don't authorize just yet.

- shopper B does the same and orders 2 units, not authorization required, the order is confirmed.

- then shopper A finally authorize payment. They get redirected to the card with notifications that the payment has failed AND that it was because of a product running out of stock.

image

Stripe correctly shows the order as canceled:

image

image


This ready to go I believe!! Thank you so much @Matt-Yorkley this is going to make some hubs very happy.

**Pull request merged**